CP-12323: The real/mock devmapper share a same interface

kaifengz opened the pull request:
CP-12323: The real/mock devmapper share a same interface

Devmapper clients could select the implementations in run-time by using the first class modules.

djs55 commented:
Hi Kaifeng,

Thanks for your pull request! I'll add some comments in-line.

djs55 commented on lib/devmapper_mock.ml:
```diff
@@ -324,68 +197,101 @@ module DeviceSet = struct
+let save_file = Mock.save_file
+let load_file = Mock.load_file
+
+module type DM_intf = sig
```
I think we should adopt a common OCaml convention and:
1. Modify the _oasis to set 'Pack: true'. For example: https://github.com/mirage/mirage-block-volume/blob/14858d04202744c2f9758f1bf0170869562705ab/_oasis#L22
2. Create a file called "s.mli" to contain the signatures. Move this signature to s.mli and rename it to "DEVMAPPER", a bit like this: https://github.com/mirage/mirage-block-volume/blob/14858d04202744c2f9758f1bf0170869562705ab/lib/s.mli#L120
3. Rename the module 'Devmapper_mock' to 'Mock'
4. Rename the module 'Devmapper' to 'Linux'
5. in the .mli files for mock.ml and linux.ml, you can include S.DEVMAPPER to inherit all the bindings e.g. https://github.com/mirage/mirage-block-volume/blob/14858d04202744c2f9758f1bf0170869562705ab/lib/vg.mli#L61

Clients will be able to

```
open Devmapper.Linux
```

or

```
open Devmapper.Mock
```

In future we could have other implementations (perhaps `Devmapper.Mirage`?)

kaifengz replied:
@djs55, thanks for comments. To enable the clients to choose which implementations in run time, should the module type `S.DEVMAPPER` be exposed too? Where should the common module type be placed in common OCaml convention?

Also, as submodules (e.g., `Location`, `Striped` and `Target`) exist within `Devmapper.Linux`, I think they are common to `Devmapper.Linux` and `Devmapper.Mock`. Where should they go then?

djs55 replied:
Hi,

I think the module type should be exposed as `Devmapper.S.DEVMAPPER`. It
should work if you put the module type into 's.mli' (no s.ml file needed).
Have a look at this example:
https://github.com/mirage/ocaml-cohttp/blob/master/lib/s.mli

I had forgotten about `Location`, `Striped` and `Target`: thanks for
reminding me! I think we could separate them into separate files so they
appear as `Devmapper.Location` `Devmapper.Striped` and `Devmapper.Target`.
So we would have:
- `Devmapper.S.DEVMAPPER`: the module type
- `Devmapper.Location`: the definition of Location
- `Devmapper.Striped`: the definition of Striped
- `Devmapper.Target`: the definition of Target
- `Devmapper.Linux`: the linux implementation which implements
  `Devmapper.S.DEVMAPPER` using `libdevicemapper`
- `Devmapper.Mock`: your new mock implementation which also implements
  `Devmapper.S.DEVMAPPER`

This will be quite similar to other projects like [mirage/ocaml-cohttp],
[mirage/ocaml-vchan] and [mirage/mirage-block-volume].

Thanks!
Dave

On Fri, May 22, 2015 at 12:36 PM, Kaifeng Zhu notifications@github.com
wrote:

> In lib/devmapper_mock.ml
> https://github.com/xapi-project/camldm/pull/8#discussion_r30887990:
> 
> > +
> > +let create = Mock.create
> > +let remove = Mock.remove
> > +let reload = Mock.reload
> > +let suspend = Mock.suspend
> > +let resume = Mock.resume
> > +let mknod = Mock.mknod
> > +let stat = Mock.stat
> > +let ls = Mock.ls
> > +let clear = Mock.clear
> > +let get_persistent = Mock.get_persistent
> > +let set_persistent = Mock.set_persistent
> > +let save_file = Mock.save_file
> > +let load_file = Mock.load_file
> > +
> > +module type DM_intf = sig
> 
> @djs55 https://github.com/djs55, thanks for comments. To enable the
> clients to choose which implementations in run time, should the module type
> S.DEVMAPPER be exposed too? Where should the common module type be placed
> in common OCaml convention?
> 
> Also, as submodules (e.g., Location, Striped and Target) exist within
> Devmapper.Linux, I think they are common to Devmapper.Linux and
> Devmapper.Mock. Where should they go then?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/xapi-project/camldm/pull/8/files#r30887990.

## 

Dave Scott

djs55 commented on lib/devmapper_mock.ml:
```diff
@@ -324,68 +197,101 @@ module DeviceSet = struct
-  devices := DeviceSet.load_file path
+type dm_intf_type = DMReal | DMMock
 
+let select_dm (intf_type : dm_intf_type) : (module DM_intf) =
```
I think this function could be moved into xenvm itself. Xenvm can parse the command-line arguments and decide which version to use.

djs55 commented:
Thanks again for your pull request! Apart from these cosmetic issues, I'm happy with it.

Commit:
CP-12323: The real/mock devmapper share a same interface

Devmapper clients could select the implementations in run-time
by using the first class modules.

Signed-off-by: Kaifeng Zhu <kaifeng.zhu@citrix.com>

kaifengz commented:
Thanks @djs55. I updated this PR according to your suggestions. Would you please review again?

djs55 commented:
This looks perfect, thanks!
